include/bout/indexoffset.hxx

```diff
@@ -0,0 +1,99 @@
+#ifndef __INDEXOFFSET_H__
+#define __INDEXOFFSET_H__
+
+#include "bout/mesh.hxx"
+#include "bout/region.hxx"
+
+/// Helper class for offsetting Ind2D/Ind3D
+///
+/// Provides methods for offsetting by fixed amounts in x, y, z, as
+/// well as a generic method for offsetting by any amount in multiple
+/// directions
+///
+/// Also provides helper methods for converting Ind2D/Ind3D to x, y, z
+/// indices
+///
+/// Examples
+/// --------
+///
+///     Field3D field, result;
+///     auto index = std::begin(region);
+///     IndexOffset<Ind3D> offset(*mesh);
+///
+///     result = field[offset.yp(*index)] - field[offset.ym(*index)];
+///
+template <typename T = Ind3D>
+struct IndexOffset {
+
+  static_assert(std::is_base_of<Ind2D, T>::value || std::is_base_of<Ind3D, T>::value,
+                "IndexOffset must be templated with either Ind2D or Ind3D");
```

Comment:
Can we also include `IndPerp` in this?  Is it as simple as checking `ny=1` in the same way as `Ind2D` does with `nz=1`?